### PR TITLE
Remote messaging - persist message (3)

### PR DESCRIPTION
**Pull request by cmonfortep**
Task/Issue URL: https://app.asana.com/0/72649045549333/1199214002677571/f

### Description

This PR introduces:
- Storing in a DB the active remote message
- remote message is stored as JSON into DB (it includes Moshi serialization and deserialization)
- Also applies translations when storing the message depending on user locale and translations provided in the remote config

### Steps to test this PR
- Tests are passing

### UI changes
No UI changes

### CODE REVIEW

**Comment by cmonfortep** [comment]
It seems CI is complaining about kotlin-reflect dependency. It seems related to a new dependency I added (moshi-adapters), I need to investigate this further but this can still be reviewed.

File: remote-messaging/remote-messaging-impl/src/main/java/com/duckduckgo/remote/messaging/impl/mappers/JsonRemoteMessageMapper.kt [hunk 1]
```diff
@@ -109,17 +112,29 @@ class JsonRemoteMessageMapper {
...
 
+    private fun RemoteMessage.localizeMessage(translations: Map<String, JsonContentTranslations>?): RemoteMessage {
+        if (translations == null) return this
+
+        val locale = deviceProperties.deviceLocale()
+        return translations[locale.asJsonFormat()]?.let {
+            this.copy(content = this.content.localize(it))
+        } ?: translations[locale.language]?.let {
+            this.copy(content = this.content.localize(it))
+        } ?: this
```
Comment (aitorvs) [inline remote-messaging/remote-messaging-impl/src/main/java/com/duckduckgo/remote/messaging/impl/mappers/JsonRemoteMessageMapper.kt:135]: nit: wow, it is hard to know what will be returned here :D

File: remote-messaging/remote-messaging-impl/src/main/java/com/duckduckgo/remote/messaging/impl/models/JsonRemoteMessagingConfig.kt [hunk 2]
```diff
@@ -26,7 +26,8 @@ data class JsonRemoteMessage(
     val id: String,
     val content: JsonContent?,
     val exclusionRules: List<Int>?,
-    val matchingRules: List<Int>?
+    val matchingRules: List<Int>?,
+    val translations: Map<String, JsonContentTranslations>?
```
Comment (aitorvs) [inline remote-messaging/remote-messaging-impl/src/main/java/com/duckduckgo/remote/messaging/impl/models/JsonRemoteMessagingConfig.kt:30]: Same comment as in previous PRs. As it is a json model and we're using list and map, `null` should not be necessary

File: remote-messaging/remote-messaging-store/src/main/java/com/duckduckgo/remote/messaging/store/RemoteMessagesDao.kt [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+    abstract fun messages(): List<RemoteMessageEntity>
+
+    @Query("select * from remote_message where status = \"DISMISSED\"")
+    abstract fun dismissedMessages(): List<RemoteMessageEntity>
+
+    @Query("update remote_message set status = :newState where id = :id")
+    abstract fun udpateState(id: String, newState: Status)
+
+    @Query("DELETE FROM remote_message WHERE status = \"SCHEDULED\"")
+    abstract fun deleteActiveMessages()
```
Comment (aitorvs) [inline remote-messaging/remote-messaging-store/src/main/java/com/duckduckgo/remote/messaging/store/RemoteMessagesDao.kt:42]: nit: should it be `deleteDismissedMessages()` (?)

File: remote-messaging/remote-messaging-impl/src/main/java/com/duckduckgo/remote/messaging/impl/mappers/JsonRemoteMessageMapper.kt (hunk 1 above)
Comment (cmonfortep) [reply remote-messaging/remote-messaging-impl/src/main/java/com/duckduckgo/remote/messaging/impl/mappers/JsonRemoteMessageMapper.kt]: I like it 💁‍♂️  but managed to simplify a bit.

File: remote-messaging/remote-messaging-impl/src/main/java/com/duckduckgo/remote/messaging/impl/models/JsonRemoteMessagingConfig.kt (hunk 2 above)
Comment (cmonfortep) [reply remote-messaging/remote-messaging-impl/src/main/java/com/duckduckgo/remote/messaging/impl/models/JsonRemoteMessagingConfig.kt]: done!

**Comment by cmonfortep** [comment]
Closing in favor of https://github.com/duckduckgo/Android/pull/1741

**Pull request closed without merging**